### PR TITLE
Add Beacon Storage

### PR DESCRIPTION
**Pull request by ogenev**
_This PR is open against the upstream `remove-rocksdb`  branch._

### What was wrong?

We need different storage mechanics for the Beacon portal network

### How was it fixed?

- create a `BeaconStorage` struct which implements the `ContentStore` trait used by the overlay service.
- create a new `lc_update` SQL table for storing and querying `LigthClientUpdate` content (see #1006).
- Add beacon specific implementations to all `ContentStore` methods.
- some common utils for `PortalStorage` and `BeaconStorage` were extracted

It may be useful to review this PR commit by commit.

All storage code and SQL queries are in a single `storage.rs` file, which is a mess. We will refactor this at a later stage.

Ideally, we want to rename `PortalStorage` to `HistoryStorage` but I prefer to do this in a separate PR because I don't want to introduce many diffs here.

In a following PR, we will resolve the TODOs and implement a storage cache for storing lc finality and optimistic updates.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### CODE REVIEW

File: portalnet/src/storage.rs [hunk 1]
```diff
@@ -744,6 +677,496 @@ impl PortalStorage {
...
+            value_size
+        ],
+    ) {
+        Ok(_) => Ok(()),
+        Err(err) => Err(err.into()),
+    }
+}
+
 // SQLite Statements
 const CREATE_QUERY: &str = "CREATE TABLE IF NOT EXISTS content_data (
```
Comment (njgheorghita) [inline portalnet/src/storage.rs:1171]: A lot of these `const` sql query variable names feel insufficiently descriptive now that you've added the more-descriptive LC table names. But, this is just a nitpick to keep in mind in upcoming refactor prs. It probably won't matter if the `BeaconStorage` and `HistoryStorage` are separated into different modules.

File: portalnet/src/storage.rs [hunk 2]
```diff
@@ -744,6 +677,496 @@ impl PortalStorage {
...
+        match conn.execute(
+            INSERT_LC_UPDATE_QUERY,
+            params![period, value, 0, value_size],
+        ) {
+            Ok(_) => Ok(()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    pub fn store(
```
Comment (njgheorghita) [inline portalnet/src/storage.rs:885]: It feels like this (& any other shared methods inside this `impl BeaconStorage {}`) should be moved to the `ContentStore` trait... or maybe a new shared trait? I'm not sure, since IIRC `ContentStore` was primarily created to enable `MemoryContentStore` and use that inside tests. Maybe, we should create a new trait that will be shared amongst all subnetwork storage types. This could also include shared metrics reporting methods. Just thinking out loud.

File: portalnet/src/storage.rs [hunk 3]
```diff
@@ -744,6 +677,496 @@ impl PortalStorage {
...
+            }
+        }?
+        .num_bytes;
+
+        self.metrics.report_content_data_storage_bytes(sum);
+
+        Ok(sum as u64)
+    }
+
+    pub fn lookup_content_key(&self, id: [u8; 32]) -> anyhow::Result<Option<Vec<u8>>> {
```
Comment (njgheorghita) [inline portalnet/src/storage.rs:1003]: Am I missing something? Is this not essentially the same as `lookup_content_value()`?

File: portalnet/src/storage.rs [hunk 4]
```diff
@@ -744,6 +677,496 @@ impl PortalStorage {
...
+    }
+
+    /// Public method for looking up a content value by its content id
+    pub fn lookup_content_value(&self, id: [u8; 32]) -> anyhow::Result<Option<Vec<u8>>> {
+        let conn = self.sql_connection_pool.get()?;
+        lookup_content_value(id, conn)?
+    }
+
+    /// Public method for looking up a  light client update value by period number
+    pub fn lookup_lc_update_value(&self, period: u64) -> anyhow::Result<Option<Vec<u8>>> {
```
Comment (njgheorghita) [inline portalnet/src/storage.rs:1050]: Also, isn't this exactly the same as `lookup_lc_update_period()`?

File: portalnet/src/storage.rs (hunk 1 above)
Comment (ogenev) [reply portalnet/src/storage.rs:1171]: Yes, the plan is to separate those per network but the goal now is to add the functionality asap and refactor afret.

File: portalnet/src/storage.rs (hunk 2 above)
Comment (ogenev) [reply portalnet/src/storage.rs:885]: "store" is not a required method because it is wrapped by `put` which is in the `ContentStore` trait and this is what is used by tests and the overlay service. I'm using it in `BeaconStorage` because I try to follow a similar structure as `PortalStorage`, but this shouldn't be a hard requirement.

Anyway, I see some benefits of having additional traits but imo we will have more visibility here when we start to refactor the code.

File: portalnet/src/storage.rs (hunk 3 above)
Comment (ogenev) [reply portalnet/src/storage.rs:1003]: `lookup_content_value` accepts content id and give us a content value result, while `lookup_content_key` accepts content id and returns a content key.

File: portalnet/src/storage.rs (hunk 4 above)
Comment (ogenev) [reply portalnet/src/storage.rs:1050]: Similar logic as above, `lookup_lc_update_period` returns light client update period number while `lookup_lc_update_value` returns the light client update ssz bytes (content value).

File: portalnet/src/storage.rs [hunk 5]
```diff
@@ -744,6 +677,496 @@ impl PortalStorage {
...
+
+    fn put<K: OverlayContentKey, V: AsRef<[u8]>>(
+        &mut self,
+        key: K,
+        value: V,
+    ) -> Result<(), ContentStoreError> {
+        self.store(&key, &value.as_ref().to_vec())
+    }
+
+    /// The "radius: concept is not applicable for Beacon network
```
Comment (njgheorghita) [inline portalnet/src/storage.rs:762]: Seems like this function name/purpose should be updated in upcoming refactoring so that it can be shared amongst storages while also being logically consistent.

File: portalnet/src/storage.rs (hunk 3 above)
Comment (njgheorghita) [reply portalnet/src/storage.rs:1003]: Ahhh, yeah that makes sense. Imo it's worth adding a docstring here (and to `lookup_content_value()`), even though the fn names are quite descriptive, it can be useful (at least I would've found it useful in this case) to have the functionality spelled out in plain english

**Review by njgheorghita — APPROVED** [review]
Looks good! Thanks for answering my questions. Agreed with all the todo refactoring ideas. 🚢
